packages/cli/src/services/project.service.ts

```diff
@@ -59,24 +65,43 @@ export class ProjectService {
 			);
 		}
 
-		// 1. disable all workflows
+		let targetProject: Project | null = null;
+		if (migrateToProject) {
+			targetProject = await this.getProjectWithScope(user, migrateToProject, [
+				'credential:create',
+				'workflow:create',
+			]);
+
+			if (!targetProject) {
+				throw new NotFoundError(
+					`Could not find project to migrate to. ID: ${targetProject}. You may lack permissions to create workflow and credentials in the target project.`,
+				);
+			}
+		}
 
+		// 1. disable all workflows (unless they are migrated)
 		const ownedSharedWorkflows = await this.sharedWorkflowRepository.find({
 			where: { projectId: project.id, role: 'workflow:owner' },
 			relations: { workflow: true },
 		});
 
-		// NOTE: This is supposed to happen outside the transaction.
-		// We'd otherwise need to pass the em through to a lot of functions even ending up in core.
-		// See: https://github.com/n8n-io/n8n/pull/8904#discussion_r1530150510
-		for (const sharedWorkflow of ownedSharedWorkflows) {
-			await activeWorkflowRunner.remove(sharedWorkflow.workflow.id);
+		if (!targetProject) {
+			// NOTE: This is supposed to happen outside the transaction.
+			// We'd otherwise need to pass the em through to a lot of functions even ending up in core.
+			// See: https://github.com/n8n-io/n8n/pull/8904#discussion_r1530150510
+			for (const sharedWorkflow of ownedSharedWorkflows) {
+				await activeWorkflowRunner.remove(sharedWorkflow.workflow.id);
+			}
 		}
 
 		await this.projectRelationRepository.manager.transaction(async (em) => {
-			// 2. delete workflows owned by this project
+			// 2. delete or migrate workflows owned by this project
 			for (const sharedWorkflow of ownedSharedWorkflows) {
-				await workflowService.deleteInactiveWorkflow(user, sharedWorkflow.workflow, em);
+				if (targetProject) {
+					await this.sharedWorkflowRepository.makeOwner(sharedWorkflow.workflow, targetProject, em);
+				} else {
+					await workflowService.deleteInactiveWorkflow(user, sharedWorkflow.workflow, em);
```

Comment:
@ivov I've seen that you wrote most of the code related to the binary data storage. This line of code here runs inside a transaction (thus will block other writes until it's done), this will also delete all binary data that the workflow produced.

Do you think deleting binary data from the file system or S3 could take long and we should avoid doing this within a transaction?